File: day-15/src/part2.zig
const std = @import("std");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();

    const input = @embedFile("input.txt");

    const out = process(allocator, input);
    std.debug.print("OUT: {d}\n", .{out});
}

fn process(allocator: std.mem.Allocator, input: []const u8) usize {
    _ = allocator;

    var sum: usize = 0;

    var left: usize = 0;
    for (input, 0..) |c, i| {
        if (c == ',') {
            const hash = holiday_string_helper(input[left..i]);
            std.debug.print("Hash for \"{s}\" is {d} ({d}, {d})\n", .{ input[left..i], hash, left, i });
            sum += hash;
            left = i + 1;
        }
    }
    const hash = holiday_string_helper(input[left..]);
    std.debug.print("Hash for \"{s}\" is {d} ({d}, {d})\n", .{ input[left..], hash, left, input.len });
    sum += hash;

    return sum;
}

fn holiday_string_helper(string: []const u8) usize {
    var hash: usize = 0;
    for (string) |c| {
        hash += c;
        hash *= 17;
        hash %= 256;
    }
    return hash;
}

test "example" {
    const input = "rn=1,cm-,qp=3,cm=2,qp-,pc=4,ot=9,ab=5,pc-,pc=6,ot=7";

    try std.testing.expectEqual(@as(usize, 1320), process(std.testing.allocator, input));
}

File: day-18/src/bin/part2.rs
use itertools::Itertools;
use nom::{
    character::complete::{self, alpha1, hex_digit1, multispace1, space1},
    multi::separated_list1,
    sequence::{delimited, tuple},
    IResult, Parser,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Direction {
    Up,
    Down,
    Left,
    Right,
}

impl TryFrom<char> for Direction {
    type Error = ();

    fn try_from(value: char) -> Result<Self, Self::Error> {
        Ok(match value {
            'U' | 'u' | '3' => Direction::Up,
            'D' | 'd' | '1' => Direction::Down,
            'L' | 'l' | '2' => Direction::Left,
            'R' | 'r' | '0' => Direction::Right,
            _ => return Err(()),
        })
    }
}

impl std::ops::Not for Direction {
    type Output = Direction;

    fn not(self) -> Self::Output {
        match self {
            Self::Down => Self::Up,
            Self::Up => Self::Down,
            Self::Left => Self::Right,
            Self::Right => Self::Left,
        }
    }
}

#[derive(Debug)]
struct Instruction {
    direction: Direction,
    steps: isize,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct Coord {
    x: isize,
    y: isize,
}

fn process(input: &str) -> String {
    let (_, instructions) = parse(input).unwrap();
    dbg!(&instructions, instructions.len());

    let mut area = 0;

    let mut inst_iter = instructions.chunks(2);

    while let Some([inst1, inst2]) = inst_iter.next() {
        let w = match inst1.direction {
            Direction::Down | Direction::Right => inst1.steps,
            Direction::Up | Direction::Left => -inst1.steps,
        };
        let h = match inst2.direction {
            Direction::Down | Direction::Right => inst2.steps,
            Direction::Up | Direction::Left => -inst2.steps,
        };
        dbg!(inst1, inst2, w, h, w * h);
        area += w * h;
    }

    // dbg!(&trench, trench.len());

    // for y in min_y..=max_y {
    //     for x in min_x..=max_x {
    //         print!(
    //             "{}",
    //             trench
    //                 .contains(&Coord { x, y })
    //                 .then_some('#')
    //                 .unwrap_or('.')
    //         );
    //     }
    //     println!("");
    // }

    (area - 1).to_string()
}

fn parse(input: &str) -> IResult<&str, Vec<Instruction>> {
    separated_list1(
        multispace1,
        delimited(
            tuple((alpha1, space1)),
            tuple((
                complete::u8,
                tuple((space1, complete::char('('), complete::char('#'))),
                hex_digit1.map(|colorcode: &str| {
                    let steps =
                        isize::from_str_radix(&colorcode[0..colorcode.len() - 1], 16).unwrap();
                    let direction = colorcode.chars().last().unwrap().try_into().unwrap();
                    Instruction { direction, steps }
                }),
            )),
            complete::char(')'),
        )
        .map(|(d, _, inst)| Instruction {
            direction: inst.direction,
            steps: d as isize,
        }),
    )
    .parse(input)
}

fn main() {
    println!("Output: {}", process(include_str!("./input.txt")));
}

#[cfg(test)]
mod test {
    use super::*;

    const EXAMPLE: &str = "R 6 (#70c710)
    D 5 (#0dc571)
    L 2 (#5713f0)
    D 2 (#d2c081)
    R 2 (#59c680)
    D 2 (#411b91)
    L 5 (#8ceee2)
    U 2 (#caa173)
    L 1 (#1b58a2)
    U 2 (#caa171)
    R 2 (#7807d2)
    U 3 (#a77fa3)
    L 2 (#015232)
    U 2 (#7a21e3)";
    const ANSWER: &str = "952408144115";

    #[test]
    fn example() {
        assert_eq!(ANSWER, process(EXAMPLE))
    }
}
